packages/xstate-solid/src/deriveServiceState.ts

```diff
@@ -0,0 +1,39 @@
+import type { AnyState } from 'xstate';
+import type { CheckSnapshot } from './types';
+
+function isState(state: any): state is AnyState {
+  return (
+    typeof state === 'object' &&
+    'value' in state &&
+    'context' in state &&
+    'event' in state &&
+    '_event' in state &&
+    'can' in state &&
+    'matches' in state
+  );
+}
+
+/**
+ * Takes in an interpreter or actor ref and returns a State object with reactive
+ * methods or if not State, the initial value passed in
+ * @param state {AnyState | unknown}
+ */
+export const deriveServiceState = <
+  StateSnapshot extends AnyState,
+  StateReturnType = CheckSnapshot<StateSnapshot>
+>(
+  state: StateSnapshot | unknown
+): StateReturnType => {
+  if (isState(state)) {
+    return {
+      ...state,
+      toJSON: state.toJSON,
+      toStrings: state.toStrings,
+      can: state.can,
+      hasTag: state.hasTag,
+      matches: state.matches
```

Comment:
q: why do we no longer have to "track" `state` properties within those calls? how the previous and the current approach is different? if this is the preferred way to deal with this - should we also adjust the implementation in `/fsm`?
https://github.com/statelyai/xstate/blob/afc7a83f9cd33af6b5480359f78f786d217bb0fd/packages/xstate-solid/src/fsm.ts#L70

Reply:
Solid tracks the variables in the function for us now instead of us having to add them explicitly into a wrapper function

Reply:
how does it even work? what `state` refers to as `this` within `state.matches`, `state.can` etc isn't reactive (to the best of my understanding)

Reply:
`this` becomes the new object we create with `state` properties that is fed into Solid's store, so the values are also reactive

Reply:
`state.matches` is bound though so it won't change `this` internally:
https://github.com/statelyai/xstate/blob/513c00459b8d6b96c3afc0321613ee24e20cb63a/packages/core/src/State.ts#L267

Reply:
Several tests cover `matches` and other event meth are all passing. I'm not sure what changed under the hood in Solid, but it's automatically tracking the correct variables for each method. @ryansolid, does this behavior seem right to you? Or should we be wrapping each function just to be safe?

Reply:
For context, we used to need this to ensure the function "reacted" to changes: 

```typescript 
{
     ...state,
     matches(...args: Parameters<StateObject['matches']>) {
        // tslint:disable-next-line:no-unused-expression
        this.value; // reads state.value to be tracked
        return state.matches(args[0] as never);
      }
}
```

But now we only need this:

```typescript
{
     ...state,
    matches: state.matches
}
```

Reply:
Currently, there is definitely an observable difference between those things in terms of how many times a reactive scope is reevaluated. However, I fail to understand the inner workings of this right now and I'm not sure what test should be written to prove that one is better than the other.

```ts
      createEffect(() => {
        console.log('here');
        if (state.can({ type: 'FOO' })) {}
      });

      onMount(() => {
        send({ type: 'INC' });
        send({ type: 'INC' });
        send({ type: 'INC' });
        setTimeout(() => {
          send({ type: 'INC' });
        });
      });
```

We get a different amount of logs when we use `.can` in the effect than when we use some `state.matches` call there. If I bind `state.can` in the same way `state.matches` is bound in the `State`'s constructor then I get the same results for both methods (as in, the effects and the said functions are evaluated the same amount of times). Bound methods are evaluated more often than unbound ones. And an **unbound** function that is proxied with tracking is called the same amount of times as the bound one without the extra tracking.

I lack the proper understanding of Solid to assess here which behavior is actually correct - I suspect that in one of those cases, some state properties might be stale but I wasn't yet able to prove it. I'd happily brainstorm this further over a call or something, if needed.

Reply:
I reverted to using the wrapper methods for now with explicit references to the Solid store values. This will be safer in the long run as it's consistent and won't be affected by any bound weirdness.

Reply:
Do we understand the difference between both approaches though? Can a test be added to „prove” that we are doing the correct thing now? 🤔

Reply:
We do have tests checking the methods are doing the correct thing, which is, only triggering effects when their dependencies change. So we do have tests for our actual goal, I don't think we need to add any additional tests than what we currently have for these

Reply:
> We do have tests checking the methods are doing the correct thing, which is, only triggering effects when their dependencies change. 

That means that an effect depending on `state.matches` should **not** reevaluate when `state.context` changes - and that's what I'm getting to.

I think this is the expected result:
```ts
it.only(`should not reevaluate a scope depending on state.matches when state.value doesn't change`, (done) => {
  interface MachineContext {
    counter: number;
  }

  const machine = createMachine<MachineContext>({
    context: {
      counter: 0
    },
    on: {
      INCREMENT: {
        actions: assign({
          counter: (ctx) => ctx.counter + 1
        })
      }
    }
  });

  const Comp = () => {
    let calls = 0;
    const [state, send] = useMachine(machine);

    createEffect(() => {
      calls++;
      state.matches('foo');
    });

    onMount(() => {
      send({ type: 'INC' });
      send({ type: 'INC' });
      send({ type: 'INC' });
      setTimeout(() => {
        send({ type: 'INC' });
        setTimeout(() => {
          send({ type: 'INC' });
          setTimeout(() => {
            expect(calls).toBe(1);
            done();
          }, 100);
        });
      });
    });

    return null;
  };

  render(() => <Comp />);
});
```

But at the moment the `actual` value here is 4 and not 1.

If I remove the `.bind` and the "matches wrapper" in `deriveServiceState` then I get the expected 1 here. However, I'm not sure if I get it for good reasons - and I'm unsure how to exactly validate it right now.

Reply:
I understand the problem better now and see the root of the problem. Every update creates a new `matches` method in `deriveServiceState` which then is updated in our `createImmutable` method. I will update `deriveServiceState` to only re-initialize these methods when it's a new service/actor.

Reply:
This is fixed now, and I added your test case to catch this in the future for useMachine, useActor, and useMachine-fsm